File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.py
def solution(numbers):
    # 1개의 이진트리로 해당 수를 표현할 수 있는지 검증
    def valid(subtree : str) -> bool:
        # 규칙: 0인 노드 아래에 1이 없어야 함 -> True
        n = len(subtree)
        mid = n // 2 # 서브트리의 루트인덱스
        root = subtree[mid]
        left = subtree[:mid]
        right = subtree[mid+1:]

        # 종료조건: 루트가 0인데 자식이 1이 있으면
        if root == '0' and ('1' in left + right):
            return False

        # 리프노드인 경우
        if n == 1: 
            return True

        # 재귀
        return valid(left) and valid(right)
        
    answer = []

    for num in numbers:
        # 1. 주어진 수 2진수로 변환
        bnum = bin(num)[2:]

        # 2. 포화 이진트리 높이 찾기(len(bnum) <= 2^k-1)
        k = 1
        while (1<<k)-1 < len(bnum):
            k += 1

        # 3. 포화 이진트리의 전체 노드 수
        full_len = (1 << k) - 1

        # 4. 2진수 앞에 0 채우기(전체 노드 수 만큼)
        padded = bnum.zfill(full_len)

        # 5. 하나의 이진트리로 해당 수를 표현할 수 있는지 검증(1 or 0)
        answer.append(1 if valid(padded) else 0)
    return answer

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
class Solution {
    public int[] solution(long[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        
        for (int i = 0; i < n; i ++) {
            // 2진 문자열로 변환
            String bnum = Long.toBinaryString(numbers[i]);
            
            // 포화 이진트리의 최소 높이 찾기
            int k = 1;
            // 2^k-1 >= bin.length() 될 때까지 증가
            while ((1<<k)-1 < bnum.length()) {
                k++;
            }
            // 포화 트리 전체 노드 수
            int fullLen = (1 << k) -1;
            
            // fullLen 길이만큼 문자열 앞에 0 채우기
            String padded = String.format("%" + fullLen + "s", bnum)
                                    .replace(' ', '0');
            
            // 유효성검사: 표현가능이면 1, 불가능이면 0
            answer[i] = isValid(padded) ? 1 : 0;
        }
        return answer;
    }
    
    private static boolean isValid(String s) {
        int len = s.length();
        int mid = len / 2; // parent 인덱스
        char parent = s.charAt(mid); // parent 값
        String left = s.substring(0, mid);
        String right = s.substring(mid+1);
        
        // 종료조건: parent가 0인데 서브트리에 1이 있으면 false
        if (parent == '0' && (left.contains("1") || right.contains("1"))) {
            return false;
        }
        
        // 리프노드인 경우
        if (len == 1) {
            return true;
        }
        
        // 재귀호출
        return isValid(left) && isValid(right);
    }
}